From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nikolai Nechaev <Nikolay_Nechaev@mail.ru>
Date: Sat, 11 Oct 2025 13:10:40 +0900
Subject: [PATCH] Implement fugitive notifications

Right-clicking notifications will now cause them to jump to the
opposite side of the screen (left and right).

The patch also contains a commented out implementation of the behavior
when they jump away from the cursor hovering them, unless the
control key is held.
---
 .../window/notifications_manager_default.cpp  | 26 ++++++++++++++++---
 .../window/notifications_manager_default.h    |  7 +++++
 2 files changed, 29 insertions(+), 4 deletions(-)

diff --git a/Telegram/SourceFiles/window/notifications_manager_default.cpp b/Telegram/SourceFiles/window/notifications_manager_default.cpp
index bd4e460..79d6339 100644
--- a/Telegram/SourceFiles/window/notifications_manager_default.cpp
+++ b/Telegram/SourceFiles/window/notifications_manager_default.cpp
@@ -47,12 +47,11 @@ https://github.com/telegramdesktop/tdesktop/blob/master/LEGAL
 namespace Window {
 namespace Notifications {
 namespace Default {
-namespace {
 
-[[nodiscard]] QPoint notificationStartPosition() {
+[[nodiscard]] QPoint Manager::notificationStartPosition() const {
 	const auto corner = Core::App().settings().notificationsCorner();
 	const auto r = NotificationDisplayRect(Core::App().activePrimaryWindow());
-	const auto isLeft = Core::Settings::IsLeftCorner(corner);
+	const auto isLeft = _isLeft;
 	const auto isTop = Core::Settings::IsTopCorner(corner);
 	const auto x = (isLeft == rtl())
 		? (r.x() + r.width() - st::notifyWidth - st::notifyDeltaX)
@@ -61,6 +60,8 @@ namespace {
 	return QPoint(x, y);
 }
 
+namespace {
+
 internal::Widget::Direction notificationShiftDirection() {
 	auto isTop = Core::Settings::IsTopCorner(Core::App().settings().notificationsCorner());
 	return isTop ? internal::Widget::Direction::Down : internal::Widget::Direction::Up;
@@ -74,6 +75,7 @@ std::unique_ptr<Manager> Create(System *system) {
 
 Manager::Manager(System *system)
 : Notifications::Manager(system)
+, _isLeft(Core::Settings::IsLeftCorner(Core::App().settings().notificationsCorner()))
 , _inputCheckTimer([=] { checkLastInput(); }) {
 	system->settingsChanged(
 	) | rpl::on_next([=](ChangeType change) {
@@ -478,6 +480,11 @@ Manager::~Manager() {
 	clearAllFast();
 }
 
+void Manager::flipNotificationSide() {
+	_isLeft = !_isLeft;
+	settingsChanged(ChangeType::Corner);
+}
+
 namespace internal {
 
 Widget::Widget(
@@ -661,6 +668,8 @@ Notification::Notification(
 , _fromScheduled(fromScheduled)
 , _close(this, st::notifyClose)
 , _reply(this, tr::lng_notification_reply(), st::defaultBoxButton) {
+	//setMouseTracking(true);
+
 	Lang::Updated(
 	) | rpl::on_next([=] {
 		refreshLang();
@@ -1219,11 +1228,20 @@ void Notification::startHiding() {
 	hideSlow();
 }
 
+/*
+void Notification::mouseMoveEvent(QMouseEvent *e) {
+	e->ignore();
+	if (!(e->modifiers() & Qt::ControlModifier)) {
+		manager()->flipNotificationSide();
+	}
+}
+*/
+
 void Notification::mousePressEvent(QMouseEvent *e) {
 	if (!_history) return;
 
 	if (e->button() == Qt::RightButton) {
-		unlinkHistoryInManager();
+		manager()->flipNotificationSide();
 	} else {
 		e->ignore();
 		manager()->notificationActivated(myId(), {
diff --git a/Telegram/SourceFiles/window/notifications_manager_default.h b/Telegram/SourceFiles/window/notifications_manager_default.h
index 7a23561..1e8c4e8 100644
--- a/Telegram/SourceFiles/window/notifications_manager_default.h
+++ b/Telegram/SourceFiles/window/notifications_manager_default.h
@@ -45,6 +45,10 @@ public:
 		return ManagerType::Default;
 	}
 
+	[[nodiscard]] QPoint notificationStartPosition() const;
+
+	void flipNotificationSide();
+
 	template <typename Method>
 	void enumerateNotifications(Method method) {
 		for (const auto &notification : _notifications) {
@@ -97,6 +101,8 @@ private:
 
 	void subscribeToSession(not_null<Main::Session*> session);
 
+	bool _isLeft;
+
 	std::vector<std::unique_ptr<Notification>> _notifications;
 	base::flat_map<
 		not_null<Main::Session*>,
@@ -247,6 +253,7 @@ protected:
 	void enterEventHook(QEnterEvent *e) override;
 	void leaveEventHook(QEvent *e) override;
 	void paintEvent(QPaintEvent *e) override;
+	//void mouseMoveEvent(QMouseEvent *e) override;
 	void mousePressEvent(QMouseEvent *e) override;
 	bool eventFilter(QObject *o, QEvent *e) override;
 
-- 
2.51.0

